Improve React Native Form and Control implementations

andrewhl opened the pull request:
This PR adds some examples for Inputs and Pickers for React Native. It also shows how to use Picker.Items in the standard native Control.Picker. I'm not certain my code examples are entirely correct, particularly as concerns the `getTextValue` function. Please review.

andrewhl commented:
@davidkpiano Writing this documentation raised an interesting point for me. Do you think we could improve the native implementation for custom components? I'm not sure my code is even correct, but if it is necessary to redefine `mapProps` perhaps we could improve that? Like, if the component passed into `Control` returns a standard React-Native component, for which we know the event mappings, then resolve those mappings automatically?

davidkpiano commented:
@andrewhl Yes, I do believe that the native wrappers can be improved. If you would like to do that in a PR, that would be excellent!

andrewhl commented:
@davidkpiano Please review these changes. I didn't really know what the optimal API would be, so I defined a Control.Custom input that accepts an additional 'delegate' prop, and delegates to the appropriate Control component. I updated the documentation to reflect this. I tested locally and it works great. This is the perfect API for standard native iOS inputs and form controls. I don't love the additional 'delegate' prop, but it works. If you have any suggestions on how that could be improved, I'm all ears.

I noticed you don't have any tests for native.js. Do you have any recommendations for how you'd like to approach testing this? It seems the likeliest starting point is extending custom-control-component-spec.js. In our React Native application we use `react-test-renderer` and jest. I see you're using mocha. Thoughts?

andrewhl commented:
@davidkpiano Hello. Any word on this PR? Feedback would be appreciated. If it's not what you're looking for, please let me know. I can modify it as necessary if you can provide guidance for what you're looking for.

Also, I was thinking the Native Form component should pick up onPress events from child Button components, because the practice of programmatically calling dispatch(actions.submit('form')) is a bit awkward. I end up having two submit methods for each form. Something like this:

```jsx
<Form
    onSubmit={this.handleSubmit}
    model="form"
>
    ...
    <Button onPress={this.submitForm}>
         <Text>Save</Text>
     </Button>
</Form>

    submitForm = () => {
        this.props.dispatch(actions.submit('deep.addPatient'));
    }

    handleSubmit = (form) => {
        this.props.onSubmit(form);
        this.props.onClose();
    }
```
I can't set a ref on the `Form` component because `NativeForm` is exported as a stateless functional component. So I'd like to improve this behaviour too.

davidkpiano commented:
Instead of `delegate="..."`, would it be better to have something like:

```jsx
import { Control } from 'react-redux-form/native';

// ... 

<Control.DatePickerIOS model="..." />
```

What do you think about that syntax?

Otherwise, this looks great!

davidkpiano commented:
One quick question, couldn't you just do this?

```jsx
<Control.TextInput component={SomethingElse} />
// .. etc.
```

which is the same as `delegate="TextInput"`. Did this not work?

As far as testing, when I first wrote this, it was difficult/impossible (?) to write tests for React Native components, but I will be switching over to Jest which might make this easier.

andrewhl commented:
@davidkpiano Thank you. Your recommendation to override the component prop worked like a charm. I've removed my Control.Custom component and updated the documentation to reflect this approach. I've left in the documentation the bit about creating a custom Control component by overriding the mapProps, as I believe that still has value (for scenarios where people do not use a custom component that resolves to a React Native component, i.e., may have different event bindings). Please let me know your thoughts and whether we can merge this in now.

davidkpiano commented on docs/guides/react-native.md:
```diff
@@ -111,37 +111,23 @@ export default Form;
+          placeholder="Last Name"
+          model=".last_name"
+          component={Input}
+          delegate="TextInput"
```
`delegate` can be removed now, right?

andrewhl replied:
Ah yes. I'll get rid of that.

davidkpiano commented:
Awesome, this can be merged as soon as CI passes. 👍
